src/MSALWrapper/TokenFetcher.cs

```diff
@@ -0,0 +1,116 @@
+
+            /// <summary>
+            /// Gets all the attempts made to authenticate.
+            /// </summary>
+            public List<AuthFlowResult> Attempts { get; init; }
+        }
+
+        /// <summary>
+        /// Run the authenticatuion process.
+        /// </summary>
+        /// <param name="logger">A <see cref="ILogger"/> to use.</param>
+        /// <param name="client">The client ID to authenticate as.</param>
+        /// <param name="tenant">The Azure tenant containing the client.</param>
+        /// <param name="scopes">The list of scopes to request access for.</param>
+        /// <param name="mode">The <see cref="AuthMode"/>. Controls which <see cref="IAuthFlow"/>s should be used.</param>
+        /// <param name="domain">The domain (account suffix) to filter cached accounts with.</param>
+        /// <param name="prompt">A prompt hint to display to the user if needed.</param>
+        /// <param name="timeout">The max <see cref="TimeSpan"/> we should spend attempting token acquisition for.</param>
+        /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
+        public static async Task<Result> AccessTokenAsync(
+            ILogger logger,
+            Guid client,
+            Guid tenant,
+            IEnumerable<string> scopes,
+            AuthMode mode,
+            string domain,
+            string prompt,
+            TimeSpan timeout)
+        {
+            var authFlows = AuthFlowFactory.Create(
+                logger: logger,
+                authMode: mode,
+                clientId: client,
+                tenantId: tenant,
+                scopes: scopes,
+                preferredDomain: domain,
+                promptHint: prompt);
+
+            List<AuthFlowResult> results = new List<AuthFlowResult>();
+            var executor = new AuthFlowExecutor(logger, authFlows, new StopwatchTracker(timeout));
+
+            // When running multiple AzureAuth processes with the same resource, client, and tenant IDs,
+            // They may prompt many times, which is annoying and unexpected.
+            // Use Mutex to ensure that only one process can access the corresponding resource at the same time.
+            string resource = string.Join(' ', scopes);
+            string lockName = $"Local\\{resource}_{client}_{tenant}";
+
+            // The first parameter 'initiallyOwned' indicates whether this lock is owned by current thread.
+            // It should be false otherwise a deadlock could occur.
+            using (Mutex mutex = new Mutex(initiallyOwned: false, name: lockName))
+            {
+                bool lockAcquired = false;
+                try
+                {
+                    // Wait for the other session to exit.
+                    lockAcquired = mutex.WaitOne(TimeSpan.FromMinutes(15)); // TODO: Inject mutex timeout?

```

Comment:
@shalinikhare27 mentioned in #218 that we can remove the `TODO` comment here and the timespan can become `const`.